File: packages/adapters/ios/AwakeEventCollector/RawEvent.swift
import Foundation

/// Flat, JS-bound mirror of `@awake-os/core` `RawNativeEvent`. Filled directly
/// from an OS signal — no derived or interpreted fields (I-02).
struct RawEvent {
    let occurredAt: Int64
    let type: String
    let sourceType: String
    let sourceId: String
    let subjectType: String
    let subjectId: String?
    let payload: [String: Any]

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "occurredAt": occurredAt,
            "type": type,
            "sourceType": sourceType,
            "sourceId": sourceId,
            "subjectType": subjectType,
            "payload": payload,
        ]
        dict["subjectId"] = subjectId as Any? ?? NSNull()
        return dict
    }

    static func screen(_ state: String, _ occurredAt: Int64) -> RawEvent {
        RawEvent(
            occurredAt: occurredAt,
            type: "ScreenStateChanged",
            sourceType: "System",
            sourceId: "ios.screen",
            subjectType: "Screen",
            subjectId: nil,
            payload: ["state": state]
        )
    }

    static func application(_ state: String, _ packageNameHash: String, _ occurredAt: Int64) -> RawEvent {
        RawEvent(
            occurredAt: occurredAt,
            type: "ApplicationStateChanged",
            sourceType: "System",
            sourceId: "ios.app-lifecycle",
            subjectType: "Application",
            subjectId: packageNameHash,
            payload: ["state": state, "packageNameHash": packageNameHash]
        )
    }
}

func nowMs() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }

File: apps/mobile/android/app/src/main/java/os/awake/collector/RawEvent.kt
package os.awake.collector

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

/**
 * Flat, JS-bound mirror of `@awake-os/core` `RawNativeEvent`. The adapter fills
 * this from an OS signal and nothing else — no derived or interpreted fields
 * (I-02). The `@awake-os/core` normalizer is the trust boundary that turns one
 * of these into a contract-valid Event.
 */
data class RawEvent(
    val occurredAt: Long,
    val type: String,
    val sourceType: String,
    val sourceId: String,
    val subjectType: String,
    val subjectId: String?,
    val payload: Map<String, Any?>,
) {
    fun toWritableMap(): WritableMap = Arguments.createMap().apply {
        putDouble("occurredAt", occurredAt.toDouble())
        putString("type", type)
        putString("sourceType", sourceType)
        putString("sourceId", sourceId)
        putString("subjectType", subjectType)
        if (subjectId != null) putString("subjectId", subjectId) else putNull("subjectId")
        putMap("payload", Arguments.makeNativeMap(payload))
    }

    companion object {
        fun screen(state: String, occurredAt: Long) = RawEvent(
            occurredAt = occurredAt,
            type = "ScreenStateChanged",
            sourceType = "System",
            sourceId = "android.screen",
            subjectType = "Screen",
            subjectId = null,
            payload = mapOf("state" to state),
        )

        fun application(state: String, packageNameHash: String, occurredAt: Long) = RawEvent(
            occurredAt = occurredAt,
            type = "ApplicationStateChanged",
            sourceType = "System",
            sourceId = "android.usage-events",
            subjectType = "Application",
            subjectId = packageNameHash,
            payload = mapOf("state" to state, "packageNameHash" to packageNameHash),
        )
    }
}
